csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -441,295 +398,479 @@ async Task StartedAsync(IServiceObserverPrx observer, string[] services)
+                    Lazy<Task>? waitTask = null;
 
-            internal ServiceInfo(string name, ServiceStatus status, string[] args)
-            {
-                Name = name;
-                Status = status;
-                Args = args;
+                    lock (_mutex)
+                    {
+                        if (_stopTask is Lazy<Task> stopTask)
+                        {
+                            Debug.Assert(_startTask == null);
+                            if (!stopTask.IsValueCreated ||
+                                stopTask.Value.IsCompletedSuccessfully ||
+                                !stopTask.Value.IsCompleted)
+                            {
+                                performTask = stopTask;
+                            }
+                            // else, stopTask is canceled or faulted and we create a new stopTask below.
+                        }
+                        else if (_startTask is Lazy<Task> startTask)
+                        {
+                            if (!startTask.IsValueCreated)
+                            {
+                                waitTask = startTask;
+                            }
+                            else if (startTask.Value.IsFaulted || startTask.Value.IsCanceled)
+                            {
+                                // Already stopped, no need to do anything.
+                                _startTask = null;
+                                _stopTask = new(Task.CompletedTask);
+                                performTask = _stopTask;
+                            }
+                            else if (!startTask.Value.IsCompletedSuccessfully)
+                            {
+                                waitTask = startTask; // we need to wait until this task is complete
+                            }
+                            // else, startTask completed successfully and we replace it with a stopTask below.
+                        }
+                        else
+                        {
+                            Debug.Assert(false);
+                            performTask = new(Task.CompletedTask);
+                        }
+
+                        if (waitTask == null && performTask == null)
+                        {
+                            _startTask = null;
+
+                            // The async/await ensures we get a Task (i.e. successful lazy initialization) even if
+                            // StopAsync throws an exception synchronously.
+                            // Note that StopAsync is not executed immediately with _mutex locked.
+                            _stopTask = new(async () => await _service.StopAsync());
+                            performTask = _stopTask;
+                        }
+                    }
+
+                    if (waitTask != null)
+                    {
+                        await waitTask.Value;
```

Comment:
same here, should we ignore the exceptions and return a completed task if it fails?

Reply:
Fixed.